### PR TITLE
sync blocks & assets on collator

### PR DESCRIPTION
**Pull request by goldenfiredo**

### CODE REVIEW

File: phost/src/chain_client.rs [hunk 1]
```diff
@@ -175,6 +175,22 @@ pub fn storage_value_key_vec(module: &str, storage_key_name: &str) -> Vec<u8> {
     key
 }
 
+fn storage_map_key_vec(module: &str, storage_item: &str, storage_item_key: &str) -> Vec<u8> {
```
Comment (h4x3rotab) [inline phost/src/chain_client.rs:197]: // Calculates the Substrate storage key prefix for a StorageMap

File: phost/src/chain_client.rs [hunk 2]
```diff
@@ -175,6 +175,22 @@ pub fn storage_value_key_vec(module: &str, storage_key_name: &str) -> Vec<u8> {
...
+	let mut key = storage_value_key_vec(module, storage_item);
+	let item_key = hex::decode(storage_item_key).unwrap();
+	let hash = twox_64(&item_key);
+	key.extend(&hash);
+	key.extend(&item_key);
+	key
+}
+
+pub async fn get_parachain_heads(client: &XtClient, paraclient: &XtClient, hash: Option<Hash>) -> Option<Vec<u8>> {
```
Comment (h4x3rotab) [inline phost/src/chain_client.rs]: Let's move it after `pub fn check_round_end_event`

File: phost/src/chain_client.rs (hunk 1 above)
Comment (goldenfiredo) [reply phost/src/chain_client.rs:197]: Done

File: phost/src/chain_client.rs (hunk 2 above)
Comment (goldenfiredo) [reply phost/src/chain_client.rs]: Done

File: phost/src/chain_client.rs [hunk 3]
```diff
@@ -136,6 +136,13 @@ pub fn check_round_end_event(
     Ok(false)
 }
 
+pub async fn get_parachain_heads(client: &XtClient, paraclient: &XtClient, hash: Option<Hash>) -> Option<Vec<u8>> {
+	let para_key = storage_value_key_vec("ParachainInfo", "ParachainId");
```
Comment (h4x3rotab) [inline phost/src/chain_client.rs:140]: // TODO: cache the ParachainId to save one request on parachain side

File: phost/src/chain_client.rs [hunk 4]
```diff
@@ -136,6 +136,13 @@ pub fn check_round_end_event(
...
+	let para_id = get_storage(&paraclient, None, StorageKey(para_key)).await.unwrap().unwrap();
+	let key = storage_map_key_vec("Paras", "Heads", &hex::encode(para_id));
+	get_storage(&client, hash, StorageKey(key)).await.unwrap()
```
Comment (h4x3rotab) [inline phost/src/chain_client.rs]: `get_storage` returns the raw storage in `Vec<u8>`. The data in the vec is `Para.Heads`, which is also a `Vec<u8>`. So we should decode it by:

```
match get_storage(&client, hash, StorageKey(key)).await {
    Ok(Some(raw)) => Some(Vec::<u8>::decode(&mut &raw))
    _ => None
}
```

File: phost/src/main.rs [hunk 5]
```diff
@@ -404,29 +431,45 @@ async fn batch_sync_block(
...
-                let r = req_dispatch_block(pr, &dispatch_batch).await?;
-                println!("  ..dispatch_block: {:?}", r);
-
-                // Update sync state
-                synced_blocks += dispatch_batch.len();
-            }
-        }
+        let para_fin_header = chain_client::get_parachain_heads(&client, &paraclient, Some(last_header_hash)).await;
+		if para_fin_header.is_some() {
+			let para_fin_hash = &para_fin_header.unwrap()[2..34];
```
Comment (h4x3rotab) [inline phost/src/main.rs]: Please check the problem in `get_parachain_heads` first. In the past, `para_fin_header` refers was the raw storage. It was a `Vec<u8>`, and the first 2 bytes encoded the size of the vector. That's why you skipped the first 2 bytes.

However, although `Para.Heads` is a `Vec<u8>`. It's actually a block header. The first 32 bytes of a block header is the `parent_hash`. So you extracted the parent hash, not the latest block hash.

To get the latest block hash, simply calculate `blake2_256(para_heads)`. You can find blake2_256 from `sp_core` (don't use sp_io!). Or you can decode it with a block header and call `hash()`

Example:
https://github.com/Phala-Network/phala-blockchain/blob/e7a71a3257622949194d2e1bfa23b136e923cac6/scripts/debug-cli/src/main.rs#L66-L70

File: phost/src/main.rs [hunk 6]
```diff
@@ -556,8 +603,14 @@ async fn bridge(args: Args) -> Result<(), Error> {
...
-    let events_decoder = client.events_decoder();
-    println!("Connected to substrate at: {}", args.substrate_ws_endpoint.clone());
+	println!("Connected to substrate at: {}", args.substrate_ws_endpoint.clone());
+
+	let paraclient = subxt::ClientBuilder::<Runtime>::new()
+		.skip_type_sizes_check()
+		.set_url(args.collator_ws_endpoint.clone())
+		.build().await?;
+	let events_decoder = paraclient.events_decoder();
+	println!("Connected to parachain collator at: {}", args.collator_ws_endpoint.clone());
```
Comment (h4x3rotab) [inline phost/src/main.rs]:
```suggestion
	println!("Connected to parachain node at: {}", args.collator_ws_endpoint.clone());
```

File: phost/src/main.rs [hunk 7]
```diff
@@ -556,8 +603,14 @@ async fn bridge(args: Args) -> Result<(), Error> {
         .skip_type_sizes_check()
         .set_url(args.substrate_ws_endpoint.clone())
         .build().await?;
-    let events_decoder = client.events_decoder();
-    println!("Connected to substrate at: {}", args.substrate_ws_endpoint.clone());
+	println!("Connected to substrate at: {}", args.substrate_ws_endpoint.clone());
```
Comment (h4x3rotab) [inline phost/src/main.rs]:
```suggestion
	println!("Connected to relay chain node at: {}", args.substrate_ws_endpoint.clone());
```

Comment (h4x3rotab) [reply phost/src/main.rs]: Also, I think this one should use ClientBuilder::\<KusamaRuntime\> (Kusama == Polkadot == Rococo). But now we only get headers and Para.Heads from it. So no bug related to address was triggered (MultiAddress vs AccountId). Let's leave a TODO for now:

// TODO: switch to Polkadot runtime

File: phost/src/main.rs [hunk 8]
```diff
@@ -724,9 +778,10 @@ async fn bridge(args: Args) -> Result<(), Error> {
 
             // Now we are idle. Let's try to sync the egress messages.
             if !args.no_write_back {
-                let mut msg_sync = msg_sync::MsgSync::new(&client, &pr, &mut signer);
+                let mut msg_sync = msg_sync::MsgSync::new(&paraclient, &pr, &mut signer);
                 msg_sync.maybe_sync_worker_egress(&mut system_seq).await?;
-                msg_sync.maybe_sync_balances_egress(&mut balance_seq).await?;
+                msg_sync.maybe_sync_balances_egress(&mut balance_seq, BALANCES).await?;
+				msg_sync.maybe_sync_balances_egress(&mut balance_seq, ASSETS).await?;
```
Comment (h4x3rotab) [inline phost/src/main.rs]: Rename `maybe_sync_balances_egress ` to `maybe_sync_contract_egress `

File: phost/src/types.rs [hunk 9]
```diff
@@ -134,6 +134,70 @@ pub struct TransferData {
     pub signature: Vec<u8>,
 }
 
+#[derive(Serialize, Deserialize, Debug)]
```
Comment (h4x3rotab) [inline phost/src/types.rs:137]: We can safely remove `Serialize, Deserialize` for these types

File: pruntime/enclave/src/contracts/assets.rs [hunk 10]
```diff
@@ -29,16 +36,87 @@ pub struct AssetMetadataBalance {
...
 #[derive(Serialize, Deserialize, Debug, Clone)]
+#[derive(Encode, Decode)]
+pub struct TransferToken {
+	token_id: AssetId,
+	dest: AccountIdWrapper,
+	amount: chain::Balance,
+	sequence: SequenceType,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
```
Comment (h4x3rotab) [inline pruntime/enclave/src/contracts/assets.rs:47]: Safely remove all the `Serialize, Deserialize`

File: pruntime/enclave/src/lib.rs [hunk 11]
```diff
@@ -990,7 +988,10 @@ fn init_runtime(input: InitRuntimeReq) -> Result<Value, Value> {
...
...VzdGF0aW9uIFJlcG9ydCBTaWduaW5nMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAqXot4OZuphR8nudFrAFiaGxxkgma/Es/BA+tbeCTUR106AL1ENcWA4FX3K+E9BBL0/7X5rj5nIgX/R/1ubhkKWw9gfqPG3KeAtIdcv/uTO1yXv50vqaPvE1CRChvzdS/ZEBqQ5oVvLTPZ3VEicQjlytKgN9cLnxbwtuvLUK7eyRPfJW/ksddOzP8VBBniolYnRCD2jrMRZ8nBM2ZWYwnXnwYeOAHV+W9tOhAImwRwKF/95yAsVwd21ryHMJBcGH70qLagZ7Ttyt++qO/6+KAXJuKwZqjRlEtSEz8gZQeFfVYgcwSfo96oSMAzVr7V0L6HSDLRnpb6xxmbPdqNol4tQIDAQABo4GkMIGhMB8GA1UdIwQYMBaAFHhDe3amfrzQr35CN+s1fDuHAVE8MA4GA1UdDwEB/wQEAwIGwDAMBgNVHRMBAf8EAjAAMGAGA1UdHwRZMFcwVaBToFGGT2h0dHA6Ly90cnVzdGVkc2VydmljZXMuaW50ZWwuY29tL2NvbnRlbnQvQ1JML1NHWC9BdHRlc3RhdGlvblJlcG9ydFNpZ25pbmdDQS5jcmwwDQYJKoZIhvcNAQELBQADggGBAGcIthtcK9IVRz4rRq+ZKE+7k50/OxUsmW8aavOzKb0iCx07YQ9rzi5nU73tME2yGRLzhSViFs/LpFa9lpQL6JL1aQwmDR74TxYGBAIi5f4I5TJoCCEqRHz91kpG6Uvyn2tLmnIdJbPE4vYvWLrtXXfFBSSPD4Afn7+3/XUggAlc7oCTizOfbbtOFlYA4g5KcYgS1J2ZAeMQqbUdZseZCcaZZZn65tdqee8UXZlDvx0+NdO0LR+5pFy+juM0wWbu59MvzcmTXbjsi7HY6zd53Yq5K244fwFHRQ8eOB0IWB+4PfM7FeAApZvlfqlKOlLcZL2uyVmzRkyR5yW72uo9mehX44CiPJ2fse9Y6eQtcfEhMPkmHXI01sN+KwPbpA39+xOsStjhP9N1Y1a2tQAVo+yVgLgV2Hws73Fc0o3wC78qPEA+v2aRs/Be3ZFDgDyghc/1fgU+7C+P6kbqd4poyb6IW8KCJbxfMJvkordNOgOUUxndPHEi/tb/U7uLjLOgPA==".to_string();
```
Comment (h4x3rotab) [inline pruntime/enclave/src/lib.rs]: What's this?

File: pruntime/enclave/src/lib.rs [hunk 12]
```diff
@@ -1079,15 +1081,15 @@ fn parse_block(data: &Vec<u8>) -> Result<chain::SignedBlock, parity_scale_codec:
...
-    hex::encode_hex_compact(addr.as_ref())
+     match addr {
+         chain::Address::Id(id) => hex::encode_hex_compact(id.as_ref()),
+         chain::Address::Index(index) => format!("index:{:?}", index),
+         // TODO: Verify these
+         chain::Address::Raw(address) => hex::encode_hex_compact(address.as_ref()),
+         chain::Address::Address32(address) => hex::encode_hex_compact(address.as_ref()),
+         chain::Address::Address20(address) => hex::encode_hex_compact(address.as_ref())
+    }
+    //hex::encode_hex_compact(addr.as_ref())
```
Comment (h4x3rotab) [inline pruntime/enclave/src/lib.rs]: Remove the comment

File: phost/src/main.rs [hunk 13]
```diff
@@ -404,29 +431,45 @@ async fn batch_sync_block(
...
-                    .collect();
-                maybe_take_worker_snapshot(client, events_decoder, &mut dispatch_batch).await?;
-                let r = req_dispatch_block(pr, &dispatch_batch).await?;
-                println!("  ..dispatch_block: {:?}", r);
-
-                // Update sync state
-                synced_blocks += dispatch_batch.len();
-            }
-        }
+        let para_fin_header = chain_client::get_parachain_heads(&client, &paraclient, Some(last_header_hash)).await;
```
Comment (h4x3rotab) [inline phost/src/main.rs]: // TODO: pass header to TEE with storage proof to validate

File: pruntime/enclave/src/lib.rs [hunk 14]
```diff
@@ -1261,26 +1258,24 @@ fn dispatch_block(input: DispatchBlockReq) -> Result<Value, Value> {
...
-        if block.block_header.hash() != *expected_hash {
-            return Err(error_msg("Unexpected block hash"))
-        }
-        // TODO: examine extrinsic merkle tree
-    }
 
     let ecdh_privkey = ecdh::clone_key(
         local_state.ecdh_private_key.as_ref().expect("ECDH not initizlied"));
     let mut last_block = 0;
+	let mut current_bn = first_block.block_header.number - 1;
```
Comment (h4x3rotab) [inline pruntime/enclave/src/lib.rs:1261]: // TODO: need to fix #9

File: phost/src/chain_client.rs (hunk 3 above)
Comment (goldenfiredo) [reply phost/src/chain_client.rs:140]: done

File: phost/src/chain_client.rs (hunk 4 above)
Comment (goldenfiredo) [reply phost/src/chain_client.rs]: i'm not sure

File: phost/src/main.rs (hunk 5 above)
Comment (goldenfiredo) [reply phost/src/main.rs]: It worked

File: phost/src/main.rs (hunk 6 above)
Comment (goldenfiredo) [reply phost/src/main.rs]: done

File: phost/src/main.rs (hunk 7 above)
Comment (goldenfiredo) [reply phost/src/main.rs]: done

File: phost/src/main.rs (hunk 8 above)
Comment (goldenfiredo) [reply phost/src/main.rs]: fixed

File: phost/src/types.rs (hunk 9 above)
Comment (goldenfiredo) [reply phost/src/types.rs:137]: 其它合约都是这么写的，有空一起refctor

File: pruntime/enclave/src/contracts/assets.rs (hunk 10 above)
Comment (goldenfiredo) [reply pruntime/enclave/src/contracts/assets.rs:47]: 其它合约都是这么写的，有空一起refctor

File: pruntime/enclave/src/lib.rs (hunk 11 above)
Comment (goldenfiredo) [reply pruntime/enclave/src/lib.rs]: fixed

File: pruntime/enclave/src/lib.rs (hunk 12 above)
Comment (goldenfiredo) [reply pruntime/enclave/src/lib.rs]: 之前上面的注释掉了。我只是切换一下

File: phost/src/main.rs (hunk 13 above)
Comment (goldenfiredo) [reply phost/src/main.rs]: ？

File: pruntime/enclave/src/lib.rs (hunk 14 above)
Comment (h4x3rotab) [reply pruntime/enclave/src/lib.rs:1261]: Please add //TODO here

File: phost/src/chain_client.rs (hunk 3 above)
Comment (h4x3rotab) [reply phost/src/chain_client.rs:140]: Please add //TODO to the code

Comment (h4x3rotab) [reply phost/src/chain_client.rs:140]: Oh, sorry. It's already cached. Closing this comment

File: phost/src/chain_client.rs [hunk 15]
```diff
@@ -136,6 +136,24 @@ pub fn check_round_end_event(
...
+			None
+		}
+	}
+}
+
+pub async fn get_parachain_heads(client: &XtClient, hash: Option<Hash>, storage_key: StorageKey
+) -> Result<Option<Vec<u8>>, Error> {
+	get_storage(&client, hash, storage_key).await
+}
+
```
Comment (h4x3rotab) [inline phost/src/chain_client.rs:156]:
```suggestion
pub async fn get_parachain_heads(client: &XtClient, hash: Option<Hash>, storage_key: StorageKey
) -> Result<Option<Vec<u8>>, Error> {
	let raw_data = get_storage(&client, hash, storage_key).await?;
	Ok(raw_data.map(|data| Vec<u8>::decode(&mut &data)));
}
```

File: phost/src/main.rs [hunk 16]
```diff
@@ -404,29 +432,48 @@ async fn batch_sync_block(
...
-                println!("  ..dispatch_block: {:?}", r);
-
-                // Update sync state
-                synced_blocks += dispatch_batch.len();
-            }
-        }
+        let para_fin_header_data = chain_client::get_parachain_heads(&client,
+			Some(last_header_hash), paraid_storage_key.clone()).await?;
+		let para_fin_header = sp_runtime::generic::Header::<u128, sp_runtime::traits::BlakeTwo256>::decode(
+			&mut &para_fin_header_data.unwrap()[2..]);
```
Comment (h4x3rotab) [inline phost/src/main.rs:438]:
```suggestion
			&mut &para_fin_header_data.except("No head data"));
```

**Pull request merged**